ducktape/cluster/finite_subcluster.py

```diff
@@ -28,7 +28,16 @@ def __init__(self, nodes):
         self._in_use_nodes = NodeContainer()
 
     def do_alloc(self, cluster_spec):
-        allocated = self._available_nodes.remove_spec(cluster_spec)
+        # there cannot be any bad nodes here,
+        # since FiniteSubcluster operates on ClusterNode objects,
+        # which are not checked for health by NodeContainer.remove_spec
+        # however there could be an error, specifically if a test decides to alloc more nodes than are available
+        # in a previous ducktape version this exception was raised by remove_spec
+        # in this one, for consistency, we let the cluster itself deal with allocation errors
+        allocated, bad, err = self._available_nodes.remove_spec(cluster_spec)
+        if err:
+            raise InsufficientResourcesError("Not enough nodes available to allocate. " + err)
```

Comment:
man not a huge fan of this go-like error handling here. Why dont we do error handling the more pythonic way here?

Reply:
not a fan as well, but that's how it worked before too in node_container (check can_remove_spec method). The problem is that outside of a larger refactor, I don't see a good way of returning multiple types of information here, cause there are multiple outcomes of this method. I guess I could wrap this in an AllocationResult object, which would have 'success' field set to True or False, I'll try that.

There's a more fundamental issue here of using exceptions for control flow, which is not a good pattern in general, but we do use it since I didn't want to modify the signature of alloc method on the cluster. Like this method should not raise an exception when there are not enough nodes, this is  one of the expected conditions, not an exception, but we'd have to modify the signature of do_alloc and alloc to get rid of that.

Reply:
using exceptions for control flow is considered pretty pythonic https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html

generally try catching is preferred in python in my understanding, as its more of a do thing and see if it works language rather than the other way around.